### PR TITLE
Use whole history in case of undetermined tokenization of sequence

### PR DESCRIPTION
**Pull request by sbalandi**
Task: [CVS-157295](https://jira.devtools.intel.com/browse/CVS-157295)

- fist commit is cherry-pick from https://github.com/openvinotoolkit/openvino.genai/pull/1268 and https://github.com/openvinotoolkit/openvino.genai/pull/1361
- next commit includes applying comments from https://github.com/openvinotoolkit/openvino.genai/pull/1268 and adding usage of kv cache for LLM

### CODE REVIEW

File: src/cpp/src/utils.cpp [hunk 1]
```diff
@@ -266,6 +268,21 @@ ov::Core singleton_core() {
...
+    if (encoded_history.get_size() > tokenized_history.size())
+        return false;
+    auto encoded_history_data = encoded_history.data<int64_t>();
+    while(i < encoded_history.get_size()) {
+        if (encoded_history_data[i] != tokenized_history[i])
+            break;
+        i++;
+    }
+
+    return i == tokenized_history.size() - 1 || i == tokenized_history.size() - 2;
```
Comment (Wovchena) [inline src/cpp/src/utils.cpp:283]: Why is `i == tokenized_history.size() - 2` condition treated as same history? The last tokens are different in that case

File: src/cpp/src/llm_pipeline.cpp
Comment (Wovchena) [inline src/cpp/src/llm_pipeline.cpp:1]: The urgency is about reaching 24.6. But this PR targets `master`. If it's possible to retarget this PR, do that. If it's not possible, let's finish the review here and then port to 24.6.

File: src/cpp/src/visual_language/inputs_embedder.cpp [hunk 2]
```diff
@@ -75,9 +89,11 @@ class InputsEmbedder::IInputsEmbedder {
     virtual void finish_chat() {
         m_is_chat_conversation = false;
         m_is_cache_empty = true;
+        m_trust_encoded_history = true;
 
         m_history.clear();
         m_templated_chat_history.clear();
+        m_tokenized_chat_history.clear();
```
Comment (Wovchena) [inline src/cpp/src/visual_language/inputs_embedder.cpp:96]: Keep the order of vars modification aligned between start and finish chat. Currently `start_chat()` clears `m_tokenized_chat_history` after `m_trust_encoded_history`, but `m_tokenized_chat_history` is cleared in the end in `finish_chat()`.

File: src/cpp/src/visual_language/inputs_embedder.cpp [hunk 3]
```diff
@@ -112,19 +128,31 @@ class InputsEmbedder::IInputsEmbedder {
                 new_templated_chat_history = m_tokenizer.apply_chat_template(m_history, add_generation_prompt, chat_template_fallback);
             }
             ov::Tensor new_chat_tokens = m_tokenizer.encode(new_templated_chat_history).input_ids;
-            if (m_is_cache_empty) {
+            TokenizedInputs prev_chat_tokens = m_tokenizer.encode(
+                    m_templated_chat_history
+                );
```
Comment (Wovchena) [inline src/cpp/src/visual_language/inputs_embedder.cpp:133]:
```suggestion
                m_templated_chat_history
            );
```

File: src/cpp/src/visual_language/pipeline.cpp [hunk 4]
```diff
@@ -99,8 +99,14 @@ class ov::genai::VLMPipeline::VLMPipelineImpl {
         size_t request_id = 0;
         size_t block_size = 1; // not used
         bool enable_prefix_caching = false;
+
+        auto tokenized_chat_history = m_inputs_embedder->get_tokenized_chat_history();
         size_t history_size = m_language.get_tensor("attention_mask").get_shape().at(1);
         size_t inputs_embeds_size = inputs_embeds.get_shape().at(1);
+        // inputs_embeds contains whole history
+        if (inputs_embeds.get_shape().at(1) == tokenized_chat_history.size()) {
```
Comment (Wovchena) [inline src/cpp/src/visual_language/pipeline.cpp:107]:
```suggestion
        if (inputs_embeds_size == tokenized_chat_history.size()) {
```

File: src/cpp/src/llm_pipeline.cpp
Comment (Wovchena) [inline src/cpp/src/llm_pipeline.cpp:1]: Not for this PR, but we need tests for that change. I remember you said random miniCPM diverges. This PR probably fixes that, so your case can be used as a test for VLM. LLM needs its test as well, maybe you can provoke a random weights model to generate such tokens.

**Review by Wovchena — CHANGES_REQUESTED** [review]

File: src/cpp/src/llm_pipeline.cpp [hunk 5]
```diff
@@ -119,14 +127,31 @@ class StatefulLLMPipeline final : public LLMPipelineImplBase {
...
                     encoded_input = utils::subtract_chat_tokenized_inputs(new_chat_tokens, prev_chat_tokens);
+                } else {
+                    encoded_input = new_chat_tokens;
                 }
                 m_templated_chat_history = new_templated_chat_history;
-                m_tokenized_chat_history = new_chat_tokens;
+
+                m_tokenized_chat_history.clear();
+                std::copy(new_chat_tokens.input_ids.data<int64_t>(), new_chat_tokens.input_ids.data<int64_t>() + new_chat_tokens.input_ids.get_size(),
+                          std::back_inserter(m_tokenized_chat_history));
```
[the remark this answers is not in the document]
Comment (ilya-lavrenov) [reply src/cpp/src/llm_pipeline.cpp:153]: looks like generate with encoded inputs can always reset m_tokenized_chat_history  to true

File: src/cpp/src/llm_pipeline.cpp [hunk 6]
```diff
@@ -119,14 +127,31 @@ class StatefulLLMPipeline final : public LLMPipelineImplBase {
...
-                    encoded_input = new_chat_tokens;
-                } else {
-                    auto prev_chat_tokens = m_tokenizer.encode(m_templated_chat_history, ov::genai::add_special_tokens(false));
+                auto prev_chat_tokens = m_tokenizer.encode(m_templated_chat_history, ov::genai::add_special_tokens(false));
+
+                // some symbols combinations can be encoded by the tokenizer in different ways
+                // if we met sequence with such combination of symbols, we cannot correctly subtract the new history from the old history
+                // and find the difference as a prompt
+                // so let's check it out and use the whole history in this case
+                if (m_history_available && m_trust_encoded_history)
```
Comment (ilya-lavrenov) [inline src/cpp/src/llm_pipeline.cpp:136]: looks like `m_history_available` is redundant and we can use `!m_tokenized_chat_history.empty()`

File: src/cpp/src/llm_pipeline.cpp [hunk 7]
```diff
@@ -119,14 +127,31 @@ class StatefulLLMPipeline final : public LLMPipelineImplBase {
...
+
+                if (m_history_available && m_trust_encoded_history) {
                     encoded_input = utils::subtract_chat_tokenized_inputs(new_chat_tokens, prev_chat_tokens);
+                } else {
+                    encoded_input = new_chat_tokens;
                 }
                 m_templated_chat_history = new_templated_chat_history;
-                m_tokenized_chat_history = new_chat_tokens;
+
+                m_tokenized_chat_history.clear();
```
Comment (ilya-lavrenov) [inline src/cpp/src/llm_pipeline.cpp:151]:
```suggestion
                m_tokenized_chat_history.reserve(new_chat_tokens.input_ids.get_size());
```

File: src/cpp/src/llm_pipeline.cpp [hunk 8]
```diff
@@ -119,14 +127,31 @@ class StatefulLLMPipeline final : public LLMPipelineImplBase {
...
+                std::copy(new_chat_tokens.input_ids.data<int64_t>(), new_chat_tokens.input_ids.data<int64_t>() + new_chat_tokens.input_ids.get_size(),
```
Comment (ilya-lavrenov) [inline src/cpp/src/llm_pipeline.cpp:152]:
```suggestion
                std::copy_n(new_chat_tokens.input_ids.data<int64_t>(), new_chat_tokens.input_ids.get_size(),
```

File: src/cpp/src/llm_pipeline.cpp [hunk 9]
```diff
@@ -222,53 +258,65 @@ class StatefulLLMPipeline final : public LLMPipelineImplBase {
...
+           utils::initialize_position_ids(*position_ids, attention_mask, kv_cache_len);
         }
 
         if(m_adapter_controller) {
             m_adapter_controller->apply(m_model_runner, config.adapters);
         }
 
+        auto input_tokens = input_ids;
+        if (is_chat_conversation && !m_trust_encoded_history) {
+            input_tokens = tokenized_chat_history;
```
Comment (ilya-lavrenov) [inline src/cpp/src/llm_pipeline.cpp:305]: looks like here we can also set `m_trust_encoded_history = true;`

File: src/cpp/src/llm_pipeline.cpp (hunk 7 above)
Comment (sbalandi) [reply src/cpp/src/llm_pipeline.cpp:151]: done

File: src/cpp/src/llm_pipeline.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (sbalandi) [reply src/cpp/src/llm_pipeline.cpp:153]: added reset of `m_tokenized_chat_history` to true

File: src/cpp/src/llm_pipeline.cpp
Comment (sbalandi) [reply src/cpp/src/llm_pipeline.cpp:1]: moved https://github.com/openvinotoolkit/openvino.genai/pull/1268

File: src/cpp/src/llm_pipeline.cpp [hunk 10]
```diff
@@ -191,6 +224,9 @@ class StatefulLLMPipeline final : public LLMPipelineImplBase {
             attention_mask = data->attention_mask;
         }
 
+        if (is_chat_conversation && m_chat_input_type == ov::genai::utils::GenerationChatInputsType::ENCODED_INPUTS)
+            std::copy(input_ids.data<int64_t>(), input_ids.data<int64_t>() + input_ids.get_size(), std::back_inserter(m_tokenized_chat_history));
```
[the remark this answers is not in the document]
Comment (ilya-lavrenov) [reply src/cpp/src/llm_pipeline.cpp:228]: agree

**Comment by ilya-lavrenov** [comment]
Let's sync with https://github.com/openvinotoolkit/openvino.genai/pull/1268 and address remaining comments from that PR if they are valid

**Review by ilya-lavrenov — APPROVED** [review]

File: src/cpp/src/utils.cpp [hunk 11]
```diff
@@ -306,6 +308,51 @@ ov::Core singleton_core() {
...
+    // TODO: add handling for case with LoRA adapters enabled
+    auto states = request.query_state();
+    for (auto& state : states) {
+        if(adapter_controller && adapter_controller->has_state_name(state.get_name()))
+            continue;
+
+        ov::Tensor old_tensor = state.get_state();
+        // [BATCH_SIZE, num_kv_heads, seq_len, head_size]
+        auto shape = old_tensor.get_shape();
+        shape[2] -= remove_from_end;
```
Comment (Wovchena) [inline src/cpp/src/utils.cpp:342]: Unfortunately, sequence is represented by another dimension sometimes. See https://github.com/openvinotoolkit/openvino.genai/blob/d189eb7541a61a41581dd21361db3aa3884d211b/samples/cpp/prompt_lookup_decoding_lm/prompt_lookup_decoding_lm.cpp#L116

**Review by Wovchena — CHANGES_REQUESTED** [review]

File: src/cpp/src/utils.cpp (hunk 11 above)
Comment (sbalandi) [reply src/cpp/src/utils.cpp]: thanks for the point, added runtime search seq_len_axis

**Review by Wovchena — APPROVED** [review]

**Pull request merged**